sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/LeakDetectingSpanProcessor.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.extension.incubator.trace;
+
+import static java.lang.Thread.currentThread;
+
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.internal.shaded.WeakConcurrentMap;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import java.lang.ref.Reference;
+import java.util.Arrays;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.BiConsumer;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A {@link SpanProcessor} which will detect spans that are never ended. It will detect spans that
+ * are garbage collected without ever having `end()` called on them.
+ *
+ * <p>Note: using this SpanProcessor will definitely impact the performance of your application. It
+ * is not recommended for production use, as it uses additional memory for each span to track where
+ * a leaked span was created.
+ */
+public final class LeakDetectingSpanProcessor implements SpanProcessor {
+  private static final Logger logger = Logger.getLogger(LeakDetectingSpanProcessor.class.getName());
+
+  private final PendingSpans pendingSpans;
+
+  /**
+   * Create a new {@link LeakDetectingSpanProcessor} that will report any un-ended spans that get
+   * garbage collected.
+   */
+  public static LeakDetectingSpanProcessor create() {
+    return new LeakDetectingSpanProcessor();
+  }
+
+  private LeakDetectingSpanProcessor() {
+    this(
+        (message, throwable) ->
+            logger.log(Level.WARNING, "Span garbage collected before being ended.", throwable));
+  }
+
+  // Visible for testing
+  LeakDetectingSpanProcessor(BiConsumer<String, Throwable> reporter) {
+    pendingSpans = PendingSpans.create(reporter);
+  }
+
+  @Override
+  public void onStart(Context parentContext, ReadWriteSpan span) {
+    CallerStackTrace caller = new CallerStackTrace(span);
+    StackTraceElement[] stackTrace = caller.getStackTrace();
+
+    // take off the first 3 stack frames, as they are from the SDK itself.
+    caller.setStackTrace(
+        Arrays.copyOfRange(stackTrace, Math.min(3, stackTrace.length), stackTrace.length));
+
+    pendingSpans.put(span.getSpanContext(), caller);
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    pendingSpans.remove(span.getSpanContext()).ended = true;
+  }
+
+  @Override
+  public boolean isEndRequired() {
+    return true;
+  }
+
+  private static class PendingSpans extends WeakConcurrentMap<SpanContext, CallerStackTrace> {
```

Comment:
Just noticed - I think this should be `<Span, CallerStackTrace>`. It'd be rare but `SpanContext` could stick around longer than `Span` which is what actually matters, so may as well be precise. If we wanted to implement something like Zipkin's "flush on GC" mode in the future, we'd need to store a reference to `SpanContext` in `CallerStackTrace` and `WeakConcurrentMap<SpanContext, SomethingWithSpanContext>` will prevent GC of the spancontext forever.

Reply:
Good point. I couldn't decide what the right level to point to here. I chose spancontext because it references fewer things, but that wasn't a strong reason. 

I'll adjust it.

Reply:
Hmm. A problem with this is that `ReadableSpan` is not a `Span`. I'd have to cast to `Span` to handle `onEnd()` properly. Does that seem ok?

Reply:
I guess I can use `ReadableSpan` as the type for the reference, as the top of the interface hierarchy.

Reply:
done